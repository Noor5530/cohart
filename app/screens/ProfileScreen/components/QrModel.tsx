import React, { useState, useEffect, useCallback } from "react";
import { View, Text, StyleSheet, Image } from "react-native";
import Modal from "react-native-modal";
import { backgroundColor } from "config/colors";
import { iconColor, textColor } from "../../../config/colors";
import { appFonts } from "../../../components/text";
import {
  widthPercentageToDP,
  heightPercentageToDP,
} from "react-native-responsive-screen";
import { TouchableOpacity } from "react-native-gesture-handler";
interface Props {
  isVisible: boolean;
  toggleModal: (data: boolean) => void;
  name: string;
  link: string;
  first_name: string;
}
import RNQRGenerator from "rn-qr-generator";
import ActivityIndicator from "react-native-paper/src/components/ActivityIndicator";
import branch from "react-native-branch";
import Config from "react-native-config";

export default function QrModel({ toggleModal, isVisible, name, link, first_name }: Props) {
  const onToggleModal = () => {
    toggleModal(false);
  };
  const [qrImage, setQrImage] = useState<null | string>(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    if (isVisible) {
      generateQrCode();
    }
    return () => {
      setQrImage(null);
    };
  }, [isVisible]);

  const generateQrCode = useCallback(async () => {
    try {
      setLoading(true);
      let response = await RNQRGenerator.generate({
        value: link,
        height: widthPercentageToDP(40),
        width: widthPercentageToDP(40),
        correctionLevel: "H",
      });
      setLoading(false);
      const { uri } = response;

      setQrImage(uri);
    } catch (error) {
      setQrImage(null);
      setLoading(false);
    }
  }, []);
  
  const onClickShare = async () => {
    let branchUniversalObject = await branch.createBranchUniversalObject(
      "canonicalIdentifier",
      {
        locallyIndex: true,
        title: `Artist Profile`,
        contentDescription: `Artist Profile ${first_name}`,
        contentMetadata: {
          customMetadata: {
            username: name,
            deepLinkType: 'ArtistProfile'
          },
        },
      }
    );

    let shareOptions = {
      messageHeader: "Check this out",
      messageBody: "Check out this artist profile on Cohart!",
    };

    let linkProperties = {
      feature: "share",
      channel: "RNApp",
      tags: [name],
    };

    let controlParams;

    if (Config.BUILD === "production") {
      controlParams = {
        $desktop_url: `https://web.cohart.co/${name}`,
        $ios_url: `https://web.cohart.co/${name}`,
        $android_url: `https://web.cohart.co/${name}`,
      };
    } else {
      controlParams = {
        $desktop_url: `https://staging.cohdev.co/${name}`,
        $ios_url: `https://staging.cohdev.co/${name}`,
        $android_url: `https://staging.cohdev.co/${name}`,
      };
    }

    await branchUniversalObject.showShareSheet(
      shareOptions,
      linkProperties,
      controlParams
    );

    onToggleModal();

  }
  return (
    <Modal
      testID={"toggleModel"}
      onBackButtonPress={onToggleModal}
      onBackdropPress={onToggleModal}
      isVisible={isVisible}
      onSwipeMove={onToggleModal}
      onSwipeCancel={onToggleModal}
      deviceWidth={widthPercentageToDP(100)}
      deviceHeight={heightPercentageToDP(100)}
      propagateSwipe={true}
      style={{
        height: heightPercentageToDP(100),
      }}
    >

        <View
          style={{
            alignSelf: "center",
            width: "80%",
            backgroundColor: backgroundColor,
            paddingVertical: 25,
            borderWidth: 2,
            borderColor: iconColor,
            borderRadius: 10,
            paddingHorizontal: widthPercentageToDP(12),
            alignItems: "center",
            justifyContent: "center",
          }}
        >
          <Text style={styles.heading}>{name}</Text>
          <Text style={styles.description}>web.cohart.co/{name}</Text>
          <View style={styles.center}>
            {loading ? (
              <View
                style={{
                  height: widthPercentageToDP(50),
                  alignItems: "center",
                  justifyContent: "center",
                }}
              >
                <ActivityIndicator size={"large"} color={iconColor} />
              </View>
            ) : qrImage == null ? (
              <TouchableOpacity
                onPress={generateQrCode}
                style={{
                  height: widthPercentageToDP(50),
                  alignItems: "center",
                  justifyContent: "center",
                }}
              >
                <ActivityIndicator size={"large"} color={iconColor} />
                <Text style={styles.description}>
                  unable to load please refresh
                </Text>
              </TouchableOpacity>
            ) : (
              <Image
                style={{
                  width: widthPercentageToDP(50),
                  height: widthPercentageToDP(50),
                }}
                source={{ uri: qrImage }}
              />
            )}
          </View>
          <TouchableOpacity
            style={{
              alignItems: "center",
              justifyContent: "center",
              alignSelf: "center",
              backgroundColor: iconColor,
              width: 108,
              height: 40,
              marginTop: 5,
            }}
            onPress={onClickShare}
          >
            <Text
              style={{
                fontFamily: appFonts.InterRegular,
                fontWeight: "500",
                fontSize: 20,
                textAlign: "center",
                color: textColor,
              }}
            >
              Share
            </Text>
          </TouchableOpacity>
        </View>
    </Modal>
  );
}

const styles = StyleSheet.create({
  heading: {
    fontFamily: appFonts.InterRegular,
    fontWeight: "600",
    fontSize: 24,
    width: "80%",
    textAlign: "center",
    alignSelf: "center",
    color: iconColor,
  },
  description: {
    fontFamily: appFonts.InterRegular,
    fontWeight: "400",
    fontSize: 12,

    textAlign: "center",

    color: iconColor,
  },
  center: { alignSelf: "center" },
});
