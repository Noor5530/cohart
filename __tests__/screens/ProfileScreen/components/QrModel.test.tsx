import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import QrModel from "../../../../app/screens/ProfileScreen/components/QrModel";
import { render, fireEvent } from "@testing-library/react-native";

jest.useFakeTimers();

test("renders correctly", () => {
    let mockToggle = () => { }
    const tree = renderer.create(<QrModel toggleModal={mockToggle} isVisible={true} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test("onToggle function", () => {
    let mockToggle = () => { }
    const { getByTestId } = render(<QrModel toggleModal={mockToggle} isVisible={true} />);

    fireEvent.press(getByTestId("toggleModel"));
    expect(getByTestId("toggleModel")).not.toBeNull();
});