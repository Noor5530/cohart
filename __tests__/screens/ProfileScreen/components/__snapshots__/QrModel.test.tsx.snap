// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<mockedDefaultExport
  isVisible={true}
  onBackButtonPress={[Function]}
  onBackdropPress={[Function]}
  onSwipeCancel={[Function]}
  onSwipeMove={[Function]}
  propagateSwipe={true}
  style={
    Object {
      "height": undefined,
    }
  }
  testID="toggleModel"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#000000",
        "borderRadius": 10,
        "borderWidth": 2,
        "justifyContent": "center",
        "paddingHorizontal": undefined,
        "paddingVertical": 25,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#000000",
          "fontFamily": "Inter-Regular",
          "fontSize": 24,
          "fontWeight": "600",
          "textAlign": "center",
          "width": "80%",
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Inter-Regular",
          "fontSize": 12,
          "fontWeight": "400",
          "textAlign": "center",
        }
      }
    >
      web.cohart.co/
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": undefined,
            "justifyContent": "center",
          }
        }
      >
        <mockedDefaultExport
          color="#000000"
          size="large"
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Inter-Regular",
          "fontSize": 20,
          "fontWeight": "500",
          "textAlign": "center",
        }
      }
    >
      Share
    </Text>
  </View>
</mockedDefaultExport>
`;
